torch/csrc/autograd/python_variable.cpp

```diff
@@ -137,28 +137,49 @@ PyObject * THPVariable_Wrap(Variable var)
 static int THPVariable_traverse(THPVariable *self, visitproc visit, void *arg)
 {
   Py_VISIT(self->backward_hooks);
-  // We don't want to traverse the grad_fn, even if the Variable owns it and the
-  // shared pointer's use count is 1. This is because we would need to treat
-  // the grad_fn as part of the Python state and hold the GIL sometimes when
-  // grad_fn's shared_ptr is copied, otherwise a race condition with the Python
-  // GC could occur. Holding the GIL when the shared_ptr is copied adds
-  // undesirable complexity/overhead.
-  //
-  // When hooks, a Variable, and its grad_fn are involved in a Python reference
-  // cycle, because we're not traversing the grad_fn, the reference cycle will
-  // in fact leak.
-  //
-  // See https://gist.github.com/zou3519/7ac92b84dd7d206dcc6eae55fee8372c
-  // for more details about the race condition involving traversing the grad_fn
-  // and the python GC.
+
   const auto& tensor = THPVariable_Unpack(self);
   if (tensor.defined()) {
+
+    // WARNING: The grad_fn traversal logic is very subtle, if you change this,
+    // be very careful not to re-introduce this bug:
+    // https://gist.github.com/zou3519/7ac92b84dd7d206dcc6eae55fee8372c
+
+    // We want to be conservative here and make sure that we only traverse the
+    // grad_fn if this python object is the sole owner of the grad_fn and no
+    // other thread can create another owner without taking the GIL first.
+    // This is because we need to ensure that the "owning or not" of the grad_fn
+    // doesn't change between this traverse and the clear below.
+
+    // We ensure this is true by checking that this python object is the sole owner
+    // of the underlying Tensor and that this Tensor is the sole owner of its grad_fn.
+    // In this case, the only way to get a new reference to the grad_fn is by using
+    // this python object, which requires the GIL to be accessed.
```

Comment:
I feel like this explanation is still missing a little bit more oomph. I think what I really want to see is a general recipe for how to assess what to Py_VISIT and what not to visit, in the presence of shared pointers. Maybe it would sound something like this.

**Algorithm for determining what to visit.** Given a PyObject/C++ object that you are visiting:
1. Visit any PyObjects which you directly own (e.g., if this object were to be deallocated, you would decref that PyObject)
2. Visit any C++ object which you *uniquely* own.  That is to say, visit any `unique_ptr` references, and visit any `shared_ptr` references IFF their `use_count` is 1.

Caveat: At any point, you may conservatively choose not to visit a C++ object. When you do this, this means you may fail to break cycles which go through that C++ object; this is a good tradeoff to make if the C++ unique ownership graph may be deep and you don't want to traverse it all the way to the end.

Reply:
It also seems to me that it might be better to come up with a convention for "C++ visiting" so that you don't have to reach into other structs privates to look for Python objects. Better localization of concerns.

Reply:
Updated in the latest version. Let me know what you think!